### PR TITLE
move parse(::String, ...) to Meta module

### LINKED ISSUE #22924: Base.parse(stream::IO, ...) method is in a weird place
It's defined [in the markdown code](https://github.com/JuliaLang/julia/blob/master/base/markdown/Julia/interp.jl#L3). Seems like it would be good to move it to the same place as the rest of the parsing methods or make it a private method for the markdown code.

Also, it reads until the end of the string and then seeks back to just after the part it's actually looking for, which is a bit weird and not particularly efficient. I'm not sure the current flisp parsing code would allow us to easily write this method differently though.

### LINKED ISSUE #24349: Rename parse(::String)
I keep seeing new users (on Discourse) accidentally using e.g. `parse("1")` to go from a String to a literal. This is a few hundred times slower than `parse(Int, "1")` but in most cases, it works. I think that the function to parse source code should be named differently to avoid this. I also think it is ok if it has a slightly uglier name, e.g. `parse_sourcecode` just to be clear about what it does.

### PR DESCRIPTION
**Pull request by KristofferC**
Fix #24349.

Also fixes #22924

### CODE REVIEW

File: base/markdown/Julia/interp.jl [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-function Base.parse(stream::IO; greedy::Bool = true, raise::Bool = true)
+function Meta.parse(stream::IO; greedy::Bool = true, raise::Bool = true)
```
Comment (KristofferC) [inline base/markdown/Julia/interp.jl]: This method feels a bit out of place.

Comment (yuyichao) [reply base/markdown/Julia/interp.jl]: https://github.com/JuliaLang/julia/issues/22924

Comment (KristofferC) [reply base/markdown/Julia/interp.jl]: I'll try just making it private then.

**Comment by JeffBezanson** [comment]
Could we keep test/parse.jl in place? The purpose of it is to test language syntax, not the `Meta.parse` function specifically.

**Comment by KristofferC** [comment]
Sure.

**Comment by JeffBezanson** [comment]
Oh, that's definitely confused. Let's keep the tests for `parse(T, str)` in parse.jl, and move the rest to a new syntax.jl.

**Comment by KristofferC** [comment]
Updated.

**Pull request merged**